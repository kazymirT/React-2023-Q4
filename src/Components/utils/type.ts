export interface IForm {
  name?: string;
  age?: number;
  email?: string;
  password?: string;
  confirmPassword?: string;
  accept?: boolean;
  gender?: 'male' | 'female' | string;
  country?: string;
  picture?: unknown | File;
}

export interface ISubmitForm extends IForm {
  picture: string;
}
