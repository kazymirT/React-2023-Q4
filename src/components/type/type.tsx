import { ReactNode } from "react";

export type ComponentLoader = {
  isLoader: boolean;
};

export type ComponentDate = {
  date: ProductsType[] | null;
};

export type ChildrenContentProps = {
  data: ProductsType;
};

export type ErrorBtnProps = {
  children?: ReactNode;
};

export type ErrorBtnState = {
  hasError: boolean;
};

export type ErrorBoundaryProps = {
  children?: ReactNode;
};

export type ErrorBoundaryState = {
  hasError: boolean;
};

export type ProductResponse = {
  products: ProductsType[];
  total: string;
  skip: string;
  limit: string;
};

export type ProductsType = {
  id: number;
  title: string;
  description: string;
  price: number;
  discountPercentage: number;
  rating: number;
  stock: number;
  brand: string;
  category: string;
  thumbnail: string;
  images: [string, string, string];
};

export type SelectProps = {
  selectedValue: string;
  onChange: (event: React.ChangeEvent<HTMLSelectElement>) => void;
};

export type PaginationProps = {
  currentPage: number;
  totalPages: number;
  onPageChange: (newPage: number) => void;
};

export type PageLoader = {
  url: URL;
  data: ProductResponse;
  search: string;
  searchName: string;
};

export type DetailsLoader = {
  data: ProductsType;
  search: string;
  url: URL;
};

export type HeaderLoader = {
  searchName: string;
  search: string;
};

export type ResultsLoader = {
  data: {
    skip: string;
    products: ProductsType[];
    total: string;
    limit: string;
  };
};

export type SelectPropsType = {
  limit: string;
  onChange: (event: React.ChangeEvent<HTMLSelectElement>) => void;
};

export type PaginationPropsType = {
  limit: number;
  total: number;
  page: number;
  onChange: (pageNumber: number, totalPages: number) => void;
};

export type InitialFetchArgState = {
  limit: string;
  page: string;
  total: string;
};

export type UpdateArgPayload = {
  payload: {
    limit?: string;
    page?: string;
    total?: string;
  };
};

export type FetchArgType = {
  name: string;
  limit: string;
  page: string;
};
