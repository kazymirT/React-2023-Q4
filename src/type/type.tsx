import { ReactNode } from 'react';

export type PageProps = object;

export type PageState = {
  searchValue: string | null;
};

export type HeaderProps = {
  onClick: (value: string) => void;
};

export type SearchProps = {
  onClick: (value: string) => void;
};

export type SearchState = {
  value: string;
};

export type ContentProps = {
  data: string | null;
};

export type ComponentLoader = {
  isLoader: boolean;
};

export type ComponentDate = {
  date: ProductsType[] | null;
};

export type ChildrenContentProps = {
  data: ProductsType;
};

export type ErrorBtnProps = {
  children?: ReactNode;
};

export type ErrorBtnState = {
  hasError: boolean;
};

export type ErrorBoundaryProps = {
  children?: ReactNode;
};

export type ErrorBoundaryState = {
  hasError: boolean;
};

export type ProductResponse = {
  products: ProductsType[];
  total: number;
  skip: number;
  limit: number;
};

export type ProductsType = {
  id: number;
  title: string;
  description: string;
  price: number;
  discountPercentage: number;
  rating: number;
  stock: number;
  brand: string;
  category: string;
  thumbnail: string;
  images: [string, string, string];
};

export type SelectProps = {
  selectedValue: string;
  onChange: (event: React.ChangeEvent<HTMLSelectElement>) => void;
};

export type PaginationProps = {
  currentPage: number;
  totalPages: number;
  onPageChange: (newPage: number) => void;
};
