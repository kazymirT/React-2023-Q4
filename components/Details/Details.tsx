import styles from "./style.module.css";
import Link from "next/link";
import Image from "next/image";
import { useRouter } from "next/router";
import { updateParams } from "../utils/updateSearchParams";
import { useGetProductByIdQuery } from "../../pages/api/getData";

export const Details = () => {
  const router = useRouter();
  const params = updateParams(router.query);
  const productId: string =
    typeof router.query.id === "string" ? router.query.id : "";
  const result = useGetProductByIdQuery(productId, {
    skip: router.isFallback,
  });
  const { data } = result;

  return (
    <div className={styles.container}>
      {data && (
        <div className={styles.item}>
          <Image
            src={data.images[0]}
            className={styles.img}
            width={300}
            height={300}
            alt={data.title}
          />
          <h2>{data.title}</h2>
          <ul>
            <li>Brand: {data.brand}</li>
            <li>Category: {data.category}</li>
            <li>Price: {data.price}$</li>
          </ul>
          <p>{data.description}</p>
        </div>
      )}
      <Link
        href={params}
        data-testid="details-cancel"
        className={"details-cancel"}
      >
        Cancel
      </Link>
    </div>
  );
};
