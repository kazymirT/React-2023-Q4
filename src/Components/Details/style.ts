import styled from 'styled-components';

type DetailsContainerProps = {
  hidden: boolean;
};

export const DetailsContainer = styled.div<DetailsContainerProps>`
  display: flex;
  flex-direction: column;
  width: ${(props) => (props.hidden ? '50%' : '0%')};
`;
