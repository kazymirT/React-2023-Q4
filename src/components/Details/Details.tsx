import React from "react";
import styles from "./style.module.css";
import Link from "next/link";
import Image from "next/image";
import { useRouter } from "next/router";
import { updateParams } from "../utils/updateSearchParams";
import { ProductsType } from "../type/type";

type DetailsPropsType = {
  data: ProductsType;
};

export const Details = (props: DetailsPropsType) => {
  const router = useRouter();
  const params = updateParams(router.query);
  const { data } = props;

  return (
    <div className={styles.container} data-testid={"details"}>
      {data && (
        <div className={styles.item}>
          <Image
            src={data.images[0]}
            priority={true}
            className={styles.img}
            width={390}
            height={300}
            alt={data.title}
          />
          <h2 data-testid={"details-title"}>{data.title}</h2>
          <ul>
            <li>Brand: {data.brand}</li>
            <li>Category: {data.category}</li>
            <li>Price: {data.price}$</li>
          </ul>
          <p>{data.description}</p>
        </div>
      )}
      <Link
        href={params}
        data-testid="details-cancel"
        className={"details-cancel"}
      >
        Cancel
      </Link>
    </div>
  );
};
