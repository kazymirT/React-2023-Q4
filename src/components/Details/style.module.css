.container {
  position: relative;
  padding: 32px;
  width: 470px;
  display: flex;
  flex-direction: column;
  background-color: rgb(28, 28, 28);
  border-radius: 7px;
  height: 100%;
}
.item {
  right: 120px;
  top: 204px;
  overflow: hidden;
  max-width: 390px;
}
.img {
  border-radius: 7px;
}
.title {
  text-align: center;
  color: rgb(181, 168, 145);
}
