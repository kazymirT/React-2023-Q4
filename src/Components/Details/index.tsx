import { DetailsProps } from '../../type/type';
import { DetailsContainer } from './style';

export const Details = (props: DetailsProps) => {
  return <DetailsContainer hidden={props.aside}>Aside</DetailsContainer>;
};
