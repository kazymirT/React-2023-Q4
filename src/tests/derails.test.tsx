import React from "react";
import { describe, expect, it, vi } from "vitest";
import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import { ProductsType } from "@/components/type/type";
import { Details } from "@/components/Details/Details";
import mockRouter from "next-router-mock";
import { MemoryRouterProvider } from "next-router-mock/MemoryRouterProvider";

vi.mock("next/router", () => vi.importActual("next-router-mock"));
const data: ProductsType = {
  brand: "Apple",
  category: "smartphones",
  description: "An apple mobile which is nothing like apple",
  discountPercentage: 12.96,
  id: 1,
  images: [
    "https://i.dummyjson.com/data/products/1/3.jpg",
    "https://i.dummyjson.com/data/products/1/4.jpg",
    "https://i.dummyjson.com/data/products/1/thumbnail.jpg",
  ],
  price: 549,
  rating: 4.69,
  stock: 94,
  thumbnail: "https://i.dummyjson.com/data/products/1/thumbnail.jpg",
  title: "iPhone 9",
};

describe("Tests for the Detailed Card component", () => {
  it("Detailed card component correctly displays the detailed card data;", async () => {
    render(<Details data={data} />);
    expect(await screen.findByTestId("details-title")).toHaveTextContent(
      "iPhone 9",
    );
    expect(screen.getByText("Brand: Apple")).toBeInTheDocument();
    expect(screen.getByText("Category: smartphones")).toBeInTheDocument();
  });

  it("Clicking the close button hides the component.", async () => {
    mockRouter.push("/details/2");
    render(<Details data={data} />, { wrapper: MemoryRouterProvider });
    expect(await screen.findByTestId("details")).toBeInTheDocument();
    expect(screen.getByTestId("details")).toMatchSnapshot();
    fireEvent.click(screen.getByTestId("details-cancel"));
    expect(mockRouter.asPath).toEqual("/?limit=5&page=1&search=");
  });
});
