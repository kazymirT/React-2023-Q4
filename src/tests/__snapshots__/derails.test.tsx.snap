// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the Detailed Card component > Clicking the close button hides the component. 1`] = `
<div
  class="_container_babb22"
  data-testid="details"
>
  <div
    class="_item_babb22"
  >
    <img
      alt="iPhone 9"
      class="_img_babb22"
      data-nimg="1"
      decoding="async"
      fetchpriority="high"
      height="300"
      src="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2F3.jpg&w=828&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2F3.jpg&w=640&q=75 1x, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2F3.jpg&w=828&q=75 2x"
      style="color: transparent;"
      width="390"
    />
    <h2
      data-testid="details-title"
    >
      iPhone 9
    </h2>
    <ul>
      <li>
        Brand: 
        Apple
      </li>
      <li>
        Category: 
        smartphones
      </li>
      <li>
        Price: 
        549
        $
      </li>
    </ul>
    <p>
      An apple mobile which is nothing like apple
    </p>
  </div>
  <a
    class="details-cancel"
    data-testid="details-cancel"
    href="/?limit=5&page=1&search="
  >
    Cancel
  </a>
</div>
`;
